test/controllers/tasks_controller_test.rb

```diff
@@ -87,38 +83,129 @@
   
   # Unskip and complete these tests for Wave 3
   describe "edit" do
+    before do
+      Task.create(name: "Task #1", description: "details for the stuff to do")
+    end
+
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      id = Task.first.id
+
+      # Act
+      get edit_task_path(id)
+
+      # Assert
+      must_respond_with :success
     end
     
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      # Act
+      get task_path(-1)
+
+      # Assert
+      must_respond_with :redirect
     end
   end
   
   # Uncomment and complete these tests for Wave 3
   describe "update" do
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
+    before do
+      Task.create(name: "Task #1", description: "details for the stuff to do")
+    end
+    let (:edited_task_hash) {
+      {
+        task: {
+            name: "Edited Task",
+            description: "Edited Description",
+            completed_at: nil,
+        }
+      }
+    }
     it "can update an existing task" do
-      # Your code here
+      id = Task.first.id
+      expect {
+        patch task_path(id), params: edited_task_hash
+      }.wont_change "Task.count"
+
+      must_respond_with :redirect
+
+      task = Task.find_by(id: id)
+      expect(task.name).must_equal edited_task_hash[:task][:name]
+      expect(task.description).must_equal edited_task_hash[:task][:description]
+
     end
     
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      id = -1
+
+      expect {
+        patch task_path(id), params: edited_task_hash
+      }.wont_change "Task.count"
+
+      must_respond_with :redirect
+      must_redirect_to root_path
     end
   end
   
   # Complete these tests for Wave 4
   describe "destroy" do
     # Your tests go here
-    
+    it "can destroy an Active Record instance" do
+      # Arrange
+      task = Task.create(name: "Task #1", description: "details for the stuff to do")
+      id = task.id
+
+      # Act
+      expect {
+        delete task_path(id)
+
+        # Assert
+      }.must_change 'Task.count', -1
+
+      task = Task.find_by(name: "Task #1")
+
+      expect(task).must_be_nil
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+    end
   end
   
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+     before do
+        Task.create(name: "Task #1", description: "details for the stuff to do")
+        @task1 = Task.first
```

Comment:
Nitpick: like how you used variables above, `@task` is a perfectly great variable name; it might be better than `@task1`